include/kwk/utility/coordinates.hpp

```diff
@@ -26,30 +27,30 @@ namespace kwk
   //! @param shp  Shape used as a reference
   //! @return A tuple containing the multi-dimensional position corresponding to idx
   //================================================================================================
-  template<std::integral T, auto Desc>
-  KWK_FORCEINLINE constexpr auto coordinates(T idx, shape<Desc> const& shp) noexcept
+  template<auto Desc>
+  KWK_CONST constexpr
+  auto coordinates(std::integral auto const idx, shape<Desc> const shp) noexcept
   {
     KIWAKU_ASSERT ( idx < shp.numel()
                   ,   "Converting index " << idx
                   <<  " to an out of bounds coordinates for shape " << shp
                   );
 
-    constexpr auto nbdims = shape<Desc>::size();
-
-    // This save some 1-200ms of compile-time for those very common cases
-    if      constexpr(nbdims == 1)  return kumi::tuple{idx};
-    else if constexpr(nbdims == 2)  return kumi::tuple{idx/shp[1],idx%shp[1]};
-    else
+    /* leftmost/outer dimension does not require the modulo operation, example for 4 dimensions:
+         idx / (shp[1] * shp[2] * shp[3])          ,
+        (idx / (         shp[2] * shp[3])) % shp[1],
+        (idx /                    shp[3] ) % shp[2],
+         idx                               % shp[3],
+    */
+    return [=]<int...i>(std::integer_sequence<int, i...>)
     {
-      return  kumi::fold_right( [&](auto acc, auto m)
-                                {
-                                  auto that = kumi::push_back(acc,idx/m);
-                                  idx %= static_cast<T>(m);
-                                  return that;
-                                }
-                              , as_stride(shp)
-                              , kumi::tuple<>{}
-                              );
-    }
+      using value_type = typename shape<Desc>::value_type;
+      auto const strides{as_stride(shp)};
+      return std::array<value_type, shape<Desc>::static_order>
```

Comment:
you now return an array, which I am OK with but, the doc need to change and the tests too ;)

Reply:
i opted for the simpler (and standard) type as the result is certainly dynamic - a 'collection'/array of runtime valued integers (since idx is also a dynamic parameter) .. i.e. i don't see what extra good can a tuple bring (besides complexity;)

right about the documentation, overlooked that 😊

Reply:
The tuple was because our container accept tuple as position. I can make them accept both.

Reply:
> The tuple was because our container accept tuple as position. I can make them accept both.

🤔 i understand ... perhaps solve it at the lower level - treat arrays as kumi tuples (like std does)?

Reply:
yeah that's the way I guess

Reply:
ping :)